### PR TITLE
[8.x] Allow passing when callback to Http client retry method

### PR DESCRIPTION
**Pull request by dmason30**
**TL;DR**
A little papercut I have, this adds support for passing a `$when` callback to the `Http::retry()` method which is already supported by the `retry` helper that is used to handle the Http client retries. There are no existing tests for `Http::retry()` so this PR does not contain any.

**When Callback already exists in the retry helper**
I have just noticed whilst writing this that the `$when` callback isn't actually documented on the `retry` helper docs (https://laravel.com/docs/8.x/helpers#method-retry). It is a very useful feature where you can decide whether to continue retrying or to throw the current exception by returning a `boolean` from the callback:

```php
$response = retry(
    3, // <-- Max tries
    fn () => SomeThrowableAction::make()->execute(), // <-- Code to execute
    400, // <-- Delay between attempts
    fn ($e) => $e instanceof SomeRetryException, // <-- The when callback
 );
```

**Use with Http client**
This PR adds the `$when` support to the Http client, which can be useful if as seen in the example you only want to retry when it gets a connection failure:

```php
$response = \Http::retry(
    3,
    400,
    fn ($e) => $e instanceof ConnectionException, // <-- Added when callback support
 )->get(...);
```

### CODE REVIEW

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 1]
```diff
@@ -441,12 +448,14 @@ public function timeout(int $seconds)
      *
      * @param  int  $times
      * @param  int  $sleep
+     * @param  callable|null  $when
      * @return $this
      */
-    public function retry(int $times, int $sleep = 0)
+    public function retry(int $times, int $sleep = 0, $when = null)
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:454]: missing callable type

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Http/Client/PendingRequest.php (hunk 1 above)
Comment (dmason30) [reply src/Illuminate/Http/Client/PendingRequest.php]: added
